packages/framework/aqueduct/src/data-object-factories/dataObjectFactory.ts

```diff
@@ -55,6 +70,85 @@ export class DataObjectFactory<
 			mergedObjects.push(SharedMap.getFactory());
 		}
 
-		super(type, ctor, mergedObjects, optionalProviders, registryEntries, runtimeFactory);
+		let converted = false;
+		const convertRoundTripP = new Deferred<void>();
+
+		const fullConvertDataFn =
+			convertDataFn === undefined
+				? undefined
+				: async (runtime: FluidDataStoreRuntime) => {
+						if (!converted) {
+							converted = true;
+							submitConversionOp(runtime);
+							await convertRoundTripP.promise;
+
+							const root = (await runtime.getChannel(
+								dataObjectRootDirectoryId,
+							)) as ISharedDirectory;
+							await convertDataFn(runtime, root);
+						}
+					};
+
+		const submitConversionOp = (runtime: FluidDataStoreRuntime): void => {
+			runtime.submitMessage(DataStoreMessageType.ChannelOp, "conversion", undefined);
+		};
+
+		super(
+			type,
+			ctor,
+			mergedObjects,
+			optionalProviders,
+			registryEntries,
+			class ConverterDataStoreRuntime extends runtimeFactory {
+				public processMessages(messageCollection: IRuntimeMessageCollection): void {
+					let contents: IRuntimeMessagesContent[] = [];
+					// eslint-disable-next-line unicorn/prefer-ternary
+					if (
+						fullConvertDataFn !== undefined &&
+						// eslint-disable-next-line @typescript-eslint/no-unsafe-enum-comparison
+						messageCollection.envelope.type === DataStoreMessageType.ChannelOp
+					) {
+						if (
+							messageCollection.messagesContent.some((val) => val.contents === "conversion")
+						) {
+							convertRoundTripP.resolve();
+						}
+
+						contents = messageCollection.messagesContent.filter(
+							(val) => typeof val.contents !== "string" || val.contents !== "conversion",
```

Comment:
Curious, why test for string type? Do we have two different types of conversion op?